### PR TITLE
Float16 support in CodeGen_ARM

### PR DESCRIPTION
**Pull request by stevesuzuki-arm**
This PR is for #5449 .
Armv8-a extension of Half-precision floating point data processing is supported by CodeGen_ARM.

- The target needs to be set as 64-bit with "arm_fp16" feature.
- 32-bit is not supported in this commit.
- Upgrading fp16 to fp32 with emulated conversion is replaced with either
  - fp16 native instruction or
  - fp32 operation with native type conversion of fp16-fp32
- The dedicated unit test for NEON float16 is added.

### CODE REVIEW

**Comment by stevesuzuki-arm** [comment]
I appreciate your support! This PR is ready to review. If I'm missing is anything to move forward, please kindly let me know.

File: src/CodeGen_ARM.cpp [hunk 1]
```diff
@@ -77,6 +81,12 @@ class CodeGen_ARM : public CodeGen_Posix {
     bool neon_intrinsics_disabled() {
         return target.has_feature(Target::NoNEON);
     }
+
+    bool isFloat16AndHasFeature(const Type &t) const {
```
Comment (dsharletg) [inline src/CodeGen_ARM.cpp:85]: Halide convention would be to name this `is_float16_and_has_feature`

File: src/IROperator.cpp [hunk 2]
```diff
@@ -2179,15 +2179,29 @@ Expr fast_pow(Expr x, Expr y) {
...
-    user_assert(x.type() == Float(32)) << "fast_inverse only takes float arguments\n";
+    user_assert(x.defined()) << "fast_inverse of undefined Expr\n";
     Type t = x.type();
-    return Internal::Call::make(t, "fast_inverse_f32", {std::move(x)}, Internal::Call::PureExtern);
+    if (t == Float(32)) {
+        return Internal::Call::make(t, "fast_inverse_f32", {std::move(x)}, Internal::Call::PureExtern);
+    } else if (t == Float(16)) {
+        return Internal::Call::make(t, "fast_inverse_f16", {std::move(x)}, Internal::Call::PureExtern);
+    } else {
+        user_assert(0) << "fast_inverse only takes float arguments\n";
```
Comment (dsharletg) [inline src/IROperator.cpp:2189]: We have `user_error` instead of `user_assert(false)` for this (another one below).

File: test/correctness/float16_t_neon_op_check.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    printf("HL_TARGET is: %s\n", hl_target.to_string().c_str());
+    printf("HL_JIT_TARGET is: %s\n", jit_target.to_string().c_str());
+
+    // Only for 64bit target with fp16 feature
+    if (!(hl_target.arch == Target::ARM && hl_target.bits == 64 && hl_target.has_feature(Target::ARMFp16))) {
+        printf("[SKIP] To run this test, set HL_TARGET=arm-64-<os>-arm_fp16. \n");
+        return 0;
+    }
+    // Create Test Object
+    // Use smaller dimension than default(768, 128) to avoid fp16 overflow in reduction test case
```
Comment (dsharletg) [inline test/correctness/float16_t_neon_op_check.cpp:329]: I know this is out of scope for this PR (and Halide...), but it would be great if there were fp16 -> fp32 reduction instructions, so we could efficiently reduce fp16 inputs to fp32 outputs.

File: src/CodeGen_ARM.cpp (hunk 1 above)
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:85]: style nit: prefer snake_case for methods names in Halide, so `is_float16_and_has_feature()`

File: src/CodeGen_ARM.cpp [hunk 4]
```diff
@@ -223,8 +234,9 @@ const ArmIntrinsic intrinsic_defs[] = {
     {"vabs", "abs", UInt(16, 4), "abs", {Int(16, 4)}, ArmIntrinsic::HalfWidth},
     {"vabs", "abs", UInt(32, 2), "abs", {Int(32, 2)}, ArmIntrinsic::HalfWidth},
     {"llvm.fabs", "llvm.fabs", Float(32, 2), "abs", {Float(32, 2)}, ArmIntrinsic::HalfWidth},
+    {"llvm.fabs", "llvm.fabs", Float(16, 4), "abs", {Float(16, 4)}, ArmIntrinsic::HalfWidth | ArmIntrinsic::RequireFp16},
 
-    {"llvm.sqrt", "llvm.sqrt", Float(32, 4), "sqrt_f32", {Float(32, 4)}},
+    {"llvm.sqrt", "llvm.sqrt", Float(32, 2), "sqrt_f32", {Float(32, 2)}, ArmIntrinsic::HalfWidth},
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:239]: Is this change related to float16?

File: src/CodeGen_ARM.cpp [hunk 5]
```diff
@@ -567,6 +591,43 @@ const ArmIntrinsic intrinsic_defs[] = {
     {"vabdl_i32x2", "vabdl_i32x2", UInt(64, 2), "widening_absd", {Int(32, 2), Int(32, 2)}, ArmIntrinsic::NoMangle | ArmIntrinsic::NoPrefix},
     {"vabdl_u32x2", "vabdl_u32x2", UInt(64, 2), "widening_absd", {UInt(32, 2), UInt(32, 2)}, ArmIntrinsic::NoMangle | ArmIntrinsic::NoPrefix},
 };
+
+// List of fp16 math functions which we can avoid "emulated" equivalent code generation.
+// Only possible if the target has ARMFp16 feature.
+
+// These can be vectorized as fp16 SIMD instruction
+const std::set<string> float16_native_funcs = {
+    "sqrt_f16",
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:600]: Style nit: list the entries alphabetically

File: src/CodeGen_ARM.cpp [hunk 6]
```diff
@@ -567,6 +591,43 @@ const ArmIntrinsic intrinsic_defs[] = {
...
+    "is_nan_f16",
+    "is_inf_f16",
+    "is_finite_f16",
+};
+
+// These end up with fp32 math function call.
+// However, data type conversion of fp16 <-> fp32 is performed natively rather than emulation.
+// SIMD instruction is not available, so scalar based instruction is generated.
+const std::map<string, string> float16_transcendental_remapping = {
+    {"sin_f16", "sin_f32"},
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:614]: Style nit: list the entries alphabetically

File: src/CodeGen_ARM.cpp [hunk 7]
```diff
@@ -1316,6 +1433,11 @@ string CodeGen_ARM::mattrs() const {
             separator = ",";
         }
 
+        if (target.has_feature(Target::ARMFp16)) {
+            arch_flags += separator + "+fullfp16";
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:1437]: What minimum version of LLVM is required to support this flag?

**Review by dsharletg — COMMENTED** [review]
Thank you for this PR, great work and I'm very happy to see this support! I just have some minor nits, but I will approve the workflows to run anyways.

The only big comment I have is I wonder about splitting float16 simd_op_check into its own test. I see that the width of the test had to be reduced for reductions to work, but if we were to find a workaround for that issue, would it be better to roll this into the main simd_op_check test? That test is getting rather big and unwieldy, maybe the thing we should do is split all ARM testing into simd_op_check_arm (like we did for HVX). (But I'm fine with leaving that to a separate PR, I'm not asking you to do that split here.)

File: src/IROperator.cpp (hunk 2 above)
Comment (steven-johnson) [inline src/IROperator.cpp:2189]: Nit: Float(64) is a float argument but isn't supported, suggest something like `fast_inverse only takes float16 or float32 arguments`

File: src/IROperator.cpp [hunk 8]
```diff
@@ -2179,15 +2179,29 @@ Expr fast_pow(Expr x, Expr y) {
...
-    user_assert(x.type() == Float(32)) << "fast_inverse_sqrt only takes float arguments\n";
+    user_assert(x.defined()) << "fast_inverse_sqrt of undefined Expr\n";
     Type t = x.type();
-    return Internal::Call::make(t, "fast_inverse_sqrt_f32", {std::move(x)}, Internal::Call::PureExtern);
+    if (t == Float(32)) {
+        return Internal::Call::make(t, "fast_inverse_sqrt_f32", {std::move(x)}, Internal::Call::PureExtern);
+    } else if (t == Float(16)) {
+        return Internal::Call::make(t, "fast_inverse_sqrt_f16", {std::move(x)}, Internal::Call::PureExtern);
+    } else {
+        user_assert(0) << "fast_inverse_sqrt only takes float arguments\n";
```
Comment (steven-johnson) [inline src/IROperator.cpp:2202]: Same nit as above

File: src/runtime/arm_cpu_features.cpp [hunk 9]
```diff
@@ -22,6 +22,8 @@ WEAK CpuFeatures halide_get_cpu_features() {
 
     // TODO: add runtime detection for ARMDotProd extension
     // https://github.com/halide/Halide/issues/4727
+
+    // TODO: add runtime detection for ARMFp16 extension
```
Comment (steven-johnson) [inline src/runtime/arm_cpu_features.cpp:26]: Should probably open a tracking issue for this, too

File: src/CodeGen_ARM.cpp [hunk 10]
```diff
@@ -1097,6 +1175,24 @@ void CodeGen_ARM::visit(const Call *op) {
...
+        auto it = float16_transcendental_remapping.find(op->name);
+        if (it != float16_transcendental_remapping.end()) {
+            // This op doesn't have float16 native function.
+            // So we call float32 equivalent func with native type conversion between fp16 and fp32
+            // instead of emulated equivalent code as in EmulatedFloat16Math.cpp
+            std::vector<Expr> new_args(op->args.size());
+            for (size_t i = 0; i < op->args.size(); i++) {
+                new_args[i] = cast(Float(32, op->args[i].type().lanes()), op->args[i]);
+            }
+            auto &fp32_func_name = it->second;
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:1188]: clang-tidy complains: 

```
CodeGen_ARM.cpp:1188:13: error: 'auto &fp32_func_name' can be declared as 'const auto &fp32_func_name' [readability-qualified-auto,-warnings-as-errors]
            auto &fp32_func_name = it->second;
```

**Comment by stevesuzuki-arm** [comment]
> The only big comment I have is I wonder about splitting float16 simd_op_check into its own test

I agree with your ideas. Another point is that checking suffix for data type in vector should ideally be performed not only on floating point but also integer type test cases in simd_op_check, which I thought was too much work for this PR. So I must admit current shape is a bit compromised. As you suggested, I agree to leave that to a separate PR.

File: src/CodeGen_ARM.cpp (hunk 1 above)
Comment (stevesuzuki-arm) [reply src/CodeGen_ARM.cpp]: Done.

File: src/IROperator.cpp (hunk 2 above)
Comment (stevesuzuki-arm) [reply src/IROperator.cpp]: Done.

File: test/correctness/float16_t_neon_op_check.cpp (hunk 3 above)
Comment (stevesuzuki-arm) [reply test/correctness/float16_t_neon_op_check.cpp:329]: I agree. Thank you.

File: src/CodeGen_ARM.cpp (hunk 1 above)
Comment (stevesuzuki-arm) [reply src/CodeGen_ARM.cpp]: Done.

File: src/CodeGen_ARM.cpp (hunk 5 above)
Comment (stevesuzuki-arm) [reply src/CodeGen_ARM.cpp]: Done.

File: src/CodeGen_ARM.cpp (hunk 6 above)
Comment (stevesuzuki-arm) [reply src/CodeGen_ARM.cpp]: Done.

File: src/CodeGen_ARM.cpp (hunk 7 above)
Comment (stevesuzuki-arm) [reply src/CodeGen_ARM.cpp:1437]: I was not able to identify the exact version. But this flag is supported at least LLVM10 which Halide has already dropped support.

File: src/IROperator.cpp (hunk 2 above)
Comment (stevesuzuki-arm) [reply src/IROperator.cpp]: Done.

File: src/IROperator.cpp (hunk 8 above)
Comment (stevesuzuki-arm) [reply src/IROperator.cpp]: Done.

File: src/runtime/arm_cpu_features.cpp (hunk 9 above)
Comment (stevesuzuki-arm) [reply src/runtime/arm_cpu_features.cpp:26]: Filed #6106

File: src/CodeGen_ARM.cpp (hunk 10 above)
Comment (stevesuzuki-arm) [reply src/CodeGen_ARM.cpp]: Done.

**Comment by dsharletg** [comment]
> Another point is that checking suffix for data type in vector should ideally be performed not only on floating point but also integer type test cases in simd_op_check, which I thought was too much work for this PR.

I think we actually used to do this, but Apple's ARM assembly output doesn't have these suffixes, so we removed them.

**Review by dsharletg — APPROVED** [review]

File: test/correctness/float16_t_neon_op_check.cpp (hunk 3 above)
Comment (steven-johnson) [reply test/correctness/float16_t_neon_op_check.cpp:329]: (Let's open an issue for that so it doesn't get forgotten?)

File: src/CodeGen_ARM.cpp (hunk 4 above)
Comment (steven-johnson) [reply src/CodeGen_ARM.cpp:239]: OK, but I guess I'm confused; this looks like we now prefer to use a 2-wide sqrt_f32 instead of a 4-wide version, which seems suboptimal.

Comment (dsharletg) [reply src/CodeGen_ARM.cpp:239]: The HalfWidth flag generates the full width intrinsic too.

**Review by steven-johnson — APPROVED** [review]
LGTM, but see comment about the change to sqrt_f32 -- it was unclear to me about the change there, so I think it probably warrants a comment in the code so that future readers don't also find it unclear

**Comment by dsharletg** [comment]
I think there already is a comment in the place it should be: https://github.com/halide/Halide/blob/master/src/CodeGen_ARM.cpp#L210

That flag is used on 134 intrinsics, I don't think we should add a comment on just one of them about this.

**Comment by stevesuzuki-arm** [comment]
Many thanks for your efforts to review this patch!! Maybe I'll explore what to do for further ideas.
